### ISSUE TITLE
reduce memory consumption

### ISSUE THREAD

**Issue by vlandau** [Circuitscape/Omniscape.jl#27]
I'm getting out of memory errors using 1.2 Gb resistance surface on one worker, so I think a closer look needs to be taken at memory consumption. This is on a 32GB RAM machine, so this is no good.

One idea off the bat is to calculate inputs to 32-bit precision, but I think there are likely many more gains to be had by tweaking the code itself.

In theory, the Circuitscape problems themselves should be quite small, <<1M pixels, so most of the consumption is probably happening on the Omniscape side of things.

@ranjanan and @ViralBShah I know you're both busy, but just cc'ing you here in case you have some tips or direction on how to go about addressing this.

**Comment by ViralBShah** [comment]
Yes, with 1M pixels, Circuitscape memory usage shouldn't be an issue. Have you tried memory allocation profiling? https://docs.julialang.org/en/v1/manual/profile/

**Comment by ViralBShah** [comment]
I am going to guess that 32-bit is probably going to not give you meaningful savings.

**Comment by vlandau** [comment]
That's good to know. I'll look into profiling. I haven't tried it before. Thanks! 

@ranjan can you speak at all to the savings in memory Circuitscape got with 32 bit (or maybe that mostly done for speed savings)?

**Comment by ViralBShah** [comment]
Note that it is a bit rough and the actual allocation it reports may be off by a few lines. I still find it quite useful.

**Comment by vlandau** [comment]
Tacking this on here, noting that `readdlm` from `DelimitedFiles` uses a ton of memory (>5GB) to read in a 1.2GB .asc file and it all remains allocated.

**Comment by ViralBShah** [comment]
Probably should use something else. Can CSV.jl read these files? Or make gdal the default?

readdlm does not receive too much attention. It is not meant for such large files. More of a convenience function.

**Comment by ranjanan** [comment]
CSV.jl currently seems to error out while reading our ASC files: https://github.com/JuliaData/CSV.jl/issues/583. As for GDAL, I did an informal benchmark trying to read a 300 MB ASC file: 

```julia
julia> @time a = readdlm("../96m/cellmap.asc", skipstart=6);
 22.969700 seconds (1.92 M allocations: 4.173 GiB, 4.06% gc time)
```

and with ArchGDAL: 
```julia
julia> @time b = ArchGDAL.read(ArchGDAL.read("../96m/cellmap.asc"), 1);
 14.599338 seconds (86.79 k allocations: 370.527 MiB, 0.13% gc time)
```

We certainly need to shift away from `readdlm`. Let's wait for CSV.jl benchmark as well. It supports multithreaded reading, so lets see how much gain we get from that as well.

**Comment by vlandau** [comment]
Thanks! That is great advice. I might play around with ArchGDAL. Would I also add GDAL_jll.jl as a dependency so that GDAL gets installed (easily)?

**Comment by vlandau** [comment]
> ```julia
> julia> @time b = ArchGDAL.read(ArchGDAL.read("../96m/cellmap.asc"), 1);
> ```

@ranjanan why do you have to wrap `ArchGDAL.read` with `ArchGDAL.read` here?

**Comment by vlandau** [comment]
Alright! Got a [memory allocation profile](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem) for the functions called in Omniscape.

There appear to be some very obvious culprits.

The first one off the bat that I identified was [`clip()`](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem-L3-L19), and I've got an idea for how to significantly reduce the memory demands of that function.

Not sure what is going on [here](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem-L104), or [here](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem-L606-L607).

**Comment by ViralBShah** [comment]
Aren't those operations just creating large arrays?

**Comment by ViralBShah** [comment]
Also `clip` could easily avoid array operations using a loop,  but the garbage generated should get collected anyways. Curious to see what helps.

**Comment by ViralBShah** [comment]
What's going on with `null_current_total`?

**Comment by vlandau** [comment]
`null_current_total` has to do with an artifact correction when using "blocks" as moving window centers instead of individual pixels. I basically translated what Brad had already implemented in Omniscape.py into Julia. I'll have to revisit his code to see if there's a better way I can implement it. 

I'm wondering if GC is not working efficiently? For example, memory that was allocated during `readdlm` does not seem to be garbage collected. Switching to a function other than `readdlm` could solve that specific issue, but was just using it as an example.

**Comment by vlandau** [comment]
Also, maybe that `null_current_total` is so large because that loop caused the specific line to be evaluated 81*81 times in the run that I test on? New to memory profiling, so not sure exactly how it works :slightly_smiling_face:

**Comment by ViralBShah** [comment]
Yes, that would be because it is allocating the array each time in a loop. That's probably leading to poor performance too.

**Comment by ViralBShah** [comment]
In some cases `replace!` may work better as a way to replace values with -9999 etc. Sometimes it may be slower, but save memory. Worth trying.

**Comment by ViralBShah** [comment]
Inside the loops, depending on how small the inner loop is,  `arguments["radius"]` could be better to manually hoist outside the loop (assign to a variable outside the loop) so you don't have to repeatedly pay the cost inside the loop.

**Comment by ViralBShah** [comment]
In general, reducing memory usage may also improve performance.

**Comment by vlandau** [comment]
Thanks so much for the suggestions and insights! I'll try some of those things out.

**Comment by ViralBShah** [comment]
I am sure you have seen this. But many of those tips are related to memory usage. 

https://docs.julialang.org/en/v1/manual/performance-tips/

**Comment by vlandau** [comment]
Note to self: I think [this deepcopy](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem-L459) is entirely unnecessary.

**Comment by vlandau** [comment]
[This](https://gist.github.com/vlandau/8917ec2f23ebb3a0bc4f9cc894fde2f2#file-functions-jl-1350-mem-L606) massive memory using line is only run once before the main Omniscape program starts up in parallel, so there is some leeway.

**Comment by vlandau** [comment]
Related to #36, which was just closed by #38

**Comment by vlandau** [comment]
Omniscape is now running about 3 times faster in serial for a large problem I'm working with since applying some of these fixes, and I'm able to run it on 4 times more parallel processes! :tada:

**Comment by ViralBShah** [comment]
Close?

**Comment by vlandau** [comment]
Almost! Just checking a few more things.

**Comment by vlandau** [comment]
Actually, I might as well close. I'm thinking allowing a 32-bit option will save GBs of memory since Omniscape will in some cases be working with up to 6 separate arrays (for large problems with climate connectivity enabled), each GB's in size. Will create another issue for that though!

**Comment by ViralBShah** [comment]
Yes makes sense to do that now.
